quinn-h3/src/connection.rs

```diff
@@ -168,21 +172,58 @@ impl ConnectionInner {
                 Err(Error::UnknownStream(ty)) => println!("unknown stream type {}", ty),
                 Err(e) => {
                     self.set_error(ErrorCode::STREAM_CREATION_ERROR, format!("{:?}", e));
+                    return Err(e);
                 }
-                Ok(n) => match n {
-                    NewUni::Control(stream) => match self.recv_control {
-                        None => self.recv_control = Some(stream),
-                        Some(_) => {
-                            self.set_error(
-                                ErrorCode::STREAM_CREATION_ERROR,
-                                "control stream already open",
-                            );
-                        }
-                    },
-                    NewUni::Decoder(_) => println!("decoder stream ignored"),
-                    NewUni::Encoder(_) => println!("encoder stream ignored"),
-                    NewUni::Push(_) => println!("push stream ignored"),
-                },
+                Ok(n) => self.on_uni_resolved(n)?,
+            }
+        }
+        Ok(())
+    }
+
+    fn on_uni_resolved(&mut self, new_stream: NewUni) -> Result<(), Error> {
+        match new_stream {
+            NewUni::Control(stream) => match self.recv_control {
+                None => {
+                    self.recv_control = Some(stream);
+                    Ok(())
+                }
+                Some(_) => {
+                    self.set_error(
```

Comment:
Maybe worth coming up with some helper that prevents having to duplicate the error message twice in all these places? Looks a bit ugly.